File: src/main/java/com/company/erp/features/devices/service/DeviceService.java
package com.company.erp.features.devices.service;

import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.features.devices.dto.RegisterDeviceRequest;
import com.company.erp.features.devices.entity.Device;
import com.company.erp.features.devices.repository.DeviceRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class DeviceService {

    private final DeviceRepository devices;

    public DeviceService(DeviceRepository devices) {
        this.devices = devices;
    }

    /** Upsert by (userId, deviceId) — token rotation overwrites the row in place. */
    public Device register(Long userId, RegisterDeviceRequest req) {
        Device d = devices.findByUserIdAndDeviceId(userId, req.deviceId())
                .orElseGet(Device::new);
        d.setUserId(userId);
        d.setDeviceId(req.deviceId());
        d.setFcmToken(req.fcmToken());
        d.setPlatform(req.platform());
        d.setAppVersion(req.appVersion());
        return devices.save(d);
    }

    public void delete(Long userId, String deviceId) {
        Device d = devices.findByUserIdAndDeviceId(userId, deviceId)
                .orElseThrow(() -> new NotFoundException("Device not registered"));
        devices.delete(d);
    }

    @Transactional(readOnly = true)
    public List<Device> listForUser(Long userId) {
        return devices.findByUserId(userId);
    }

    @Transactional(readOnly = true)
    public List<Device> listForUsers(List<Long> userIds) {
        return userIds.isEmpty() ? List.of() : devices.findByUserIdIn(userIds);
    }
}

File: app/Features/Devices/Services/DeviceService.php
<?php

namespace App\Features\Devices\Services;

use App\Features\Devices\Models\Device;
use App\Support\Exceptions\NotFoundException;
use Illuminate\Support\Collection;

class DeviceService
{
    /**
     * Upsert by (userId, deviceId) — token rotation overwrites the row in place.
     *
     * @param array<string, mixed> $data  deviceId, fcmToken, platform, appVersion
     */
    public function register(int $userId, array $data): Device
    {
        $device = Device::query()
            ->where('user_id', $userId)
            ->where('device_id', $data['deviceId'])
            ->first();

        if (! $device) {
            $device = new Device();
            $device->user_id = $userId;
            $device->device_id = $data['deviceId'];
        }

        $device->fcm_token = $data['fcmToken'];
        $device->platform = $data['platform'];
        $device->app_version = $data['appVersion'] ?? null;
        $device->save();

        return $device;
    }

    public function delete(int $userId, string $deviceId): void
    {
        $device = Device::query()
            ->where('user_id', $userId)
            ->where('device_id', $deviceId)
            ->first();

        if (! $device) {
            throw new NotFoundException('Device not registered');
        }
        $device->delete();
    }

    /**
     * @return Collection<int, Device>
     */
    public function listForUser(int $userId): Collection
    {
        return Device::query()->where('user_id', $userId)->get();
    }

    /**
     * @param array<int, int> $userIds
     * @return Collection<int, Device>
     */
    public function listForUsers(array $userIds): Collection
    {
        if (empty($userIds)) {
            return collect();
        }
        return Device::query()->whereIn('user_id', $userIds)->get();
    }

    /**
     * FCM tokens for the given users.
     *
     * @param array<int, int> $userIds
     * @return array<int, string>
     */
    public function tokensForUsers(array $userIds): array
    {
        return $this->listForUsers($userIds)->pluck('fcm_token')->all();
    }
}
